### PR TITLE
[libc++] Rewrite the transitive header checking machinery

### PR DESCRIPTION
**Pull request by ldionne**
Since we don't generate a full dependency graph of headers, we can
greatly simplify the script that parses the result of --trace-includes.

At the same time, we also unify the mechanism for detecting whether a
header is a public/C compat/internal/etc header with the existing
mechanism in header_information.py.

As a drive-by this fixes the headers_in_modulemap.sh.py test which had
been disabled by mistake because it used its own way of determining
the list of libc++ headers. By consistently using header_information.py
to get that information, problems like this shouldn't happen anymore.

This should also unblock #110303, which was blocked because of
a brittle implementation of the transitive includes check which broke
when the repository was cloned at a path like /path/__something/more.

### CODE REVIEW

**Comment by ldionne** [comment]
It looks like the issue has to do with Lit's builtin `diff` command. Will have to keep investigating tomorrow.

File: libcxx/test/libcxx/transitive_includes/diff.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="""Diff two files.""",
+    )
+    parser.add_argument("file1", default=None)
+    parser.add_argument("file2", default=None)
+    args = parser.parse_args()
+
+    def doread(f):
```
Comment (EricWF) [inline libcxx/test/libcxx/transitive_includes/diff.py:22]: This should produce much better output:

```
from pathlib import Path
import argparse, difflib, sys

def main():
  parser = argparse.ArgumentParser(description='Diff two files')
  parser.add_argument('a', help='First file to diff')
  parser.add_argument('b', help='Second file to diff')
  args = parser.parse_args()

  # open two files & spray the diff to stdout; return false iff a proble
  a = Path(args.a).read_text().splitlines(keepends=True)
  b = Path(args.b).read_text().splitlines(keepends=True)
  has_diff = False
  for line in difflib.context_diff(a, b, args.a, args.b):
    has_diff = True
    print(line, end='', file=sys.stderr)

  if has_diff:
    print(f'Files differ! [{args.a}, {args.b}]', file=sys.stderr)
    sys.exit(1)


main()

```

File: libcxx/test/libcxx/transitive_includes/expected.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        choices=["cxx03", "cxx11", "cxx14", "cxx17", "cxx20", "cxx23", "cxx26"],
+    )
+    parser.add_argument(
+        "header",
+        default=None,
+        help="The header to extract the expected transitive includes for."
+    )
+    args = parser.parse_args()
+
+    CSV_ROOT = os.path.dirname(__file__)
```
Comment (EricWF) [inline libcxx/test/libcxx/transitive_includes/expected.py:30]: Alternatively `cat  cxx17.csv | grep -E "^{header} "`

Also, has anybody else noticed that these files aren't CSV?

Comment (ldionne) [reply libcxx/test/libcxx/transitive_includes/expected.py:30]: They're CSV with a space as a delimiter, which is also quite common. IDK if that's technically called CSV but it's not unseen.

File: libcxx/test/libcxx/transitive_includes/diff.py (hunk 1 above)
Comment (ldionne) [reply libcxx/test/libcxx/transitive_includes/diff.py:22]: Ack, I can change to that. TBH this whole patch is a bit throwaway, my motivation was to unblock your work on the docker-in-docker setup ASAP, and then investigate why Lit's builtin `diff` command didn't work properly.

Comment (ldionne) [reply libcxx/test/libcxx/transitive_includes/diff.py:22]: Oh, sorry actually I missed that you were using `difflib`. I was using `difflib` in an earlier version of this patch and it didn't work, so I dumbed it down to comparing the lines naively.

File: libcxx/test/libcxx/transitive_includes/expected.py (hunk 2 above)
Comment (EricWF) [reply libcxx/test/libcxx/transitive_includes/expected.py:30]: Ah, TIL.

File: libcxx/test/libcxx/transitive_includes/diff.py (hunk 1 above)
Comment (EricWF) [reply libcxx/test/libcxx/transitive_includes/diff.py:22]: I don't think difflib is at fault here, but I'm OK with this change then.

**Comment by EricWF** [comment]
So I don't think the diff tool is at fault here, because using `/usr/bin/diff` fails as well.

On my local machine, I reproduced the issue by creating a `/__w/foo/llvm-project` directory and building using that.
I suspect  there's something wrong with our escaping or regular expressions.

**Comment by ldionne** [comment]
I'll merge this to unblock a chain of dependencies getting longer and longer. I'm willing to take any post-review feedback and I can create a follow-up PR, but for now this fixes some concrete issues preventing us (amongst other things) to move to Docker-in-Docker and update the Docker images, which is blocking the clang-tidy tests re-enablement.

**Pull request merged**